### PR TITLE
Complete support for multi-source mapping in plugin mode

### PR DESCRIPTION
**Pull request by piotrtomiak**
Also, support disablement of code assist in associated non-TS files

### CODE REVIEW

File: packages/language-core/index.ts [hunk 1]
```diff
@@ -135,9 +159,22 @@ export function createLanguage<T>(
...
+							relatedMappings.forEach(mapping => allMappings.push({
+								...mapping,
+								source: relatedScriptId as T
+							}))
+						}
+					}
 					mapCache.set(
 						sourceScript.snapshot,
-						new SourceMap(virtualCode.mappings)
+						new SourceMap(allMappings, sourceScript.id)
```
Comment (piotrtomiak) [inline packages/language-core/index.ts:177]: In this mode, we need to be able to map from generated range to all of the local and associated mappings.

File: packages/source-map/lib/sourceMap.ts [hunk 2]
```diff
@@ -16,12 +16,12 @@ interface MappingMemo {
...
 }
 
-export class SourceMap<Data = unknown> {
+export class SourceMap<Data = unknown, T = unknown> {
 
 	private sourceCodeOffsetsMemo: MappingMemo | undefined;
 	private generatedCodeOffsetsMemo: MappingMemo | undefined;
 
-	constructor(public readonly mappings: Mapping<Data>[]) { }
+	constructor(public readonly mappings: ({ source?: T } & Mapping<Data>)[], private scriptId?: T) { }
```
Comment (piotrtomiak) [inline packages/source-map/lib/sourceMap.ts:24]: We need SourceMap of the generated file to be able to map from generated offset to any of the local and associated offsets

File: packages/typescript/lib/node/utils.ts [hunk 3]
```diff
@@ -1,11 +1,34 @@
...
+	if (sourceScript?.targetIds && sourceScript?.targetIds.size > 0) {
+		const sourceId = sourceScript.id
+		for (const targetId of sourceScript.targetIds) {
+			sourceScript = language.scripts.get(targetId)
+			if (sourceScript?.generated) {
+				const serviceScript = sourceScript.generated.languagePlugin.typescript?.getServiceScript(sourceScript.generated.root);
+				if (serviceScript) {
+					for (const [id, _snapshot, map] of language.maps.forEach(serviceScript.code)) {
+						if (id === sourceId) {
+							return [serviceScript, sourceScript, map] as const
```
Comment (piotrtomiak) [inline packages/typescript/lib/node/utils.ts:21]: We return here target source script to be able to perform analysis, but we keep map from the original source file to properly map from source to generated offset.

**Comment by piotrtomiak** [comment]
@johnsoncodehk - it is difficult to share the whole repo, as templates are transpiled on JVM side for performance reasons - you need to build the whole Angular application model first. Here is the plugin project: https://drive.google.com/file/d/1y7hDmtP7l1VsWWR48jxv0PngFGLPGU3X/view?usp=sharing. It is only the TypeScript side, which runs as TSC plugin. You can see that transpiled template is sent to the server with `webStormNgUpdateTranspiledTemplate` command. I guess you could mock it up, or I can prepare for you a nightly build of WebStorm, where you can check the whole code flow. You can debug the TS server when it's running within WebStorm.

**Comment by johnsoncodehk** [comment]
> Here is the plugin project: https://drive.google.com/file/d/1y7hDmtP7l1VsWWR48jxv0PngFGLPGU3X/view?usp=sharing.

Thanks! Very useful. 👍

I've released 2.3.0-alpha.8 for this, please let me know if you find problems, or please feel free to send a PR.

**Comment by johnsoncodehk** [comment]
> Update: I seem to have broken the use case of doing a find definition jump from .html to .ts and am trying to fix it.

It should be fixed by https://github.com/volarjs/volar.js/commit/775c621e0cd18dceab4575322f3936b1863c5f8a.

**Comment by piotrtomiak** [comment]
@johnsoncodehk - thanks for working on this PR and all the fixes! Really appreciated!

**Pull request merged**